ansible/files/telemetry.sh

```diff
@@ -1,19 +1,111 @@
 #!/bin/bash
 
-cp -r /home/hadoop /mnt
-sudo mount --bind --verbose /mnt/hadoop /home/hadoop
-
 # logging for any errors during bootstrapping
 exec > >(tee -i /var/log/bootstrap-script.log)
 exec 2>&1
 
 # we won't use `set -e` because that means that AWS would terminate the instance and we wouldn't get logs for why it failed
 
+# Check for master node
+IS_MASTER=true
+if [ -f /mnt/var/lib/info/instance.json ]
+then
+    IS_MASTER=$(jq .isMaster /mnt/var/lib/info/instance.json)
+fi
+
+# Parse arguments
+while [ $# -gt 0 ]; do
+    case "$1" in
+        --public-key)
+            shift
+            PUBLIC_KEY=$1
+            ;;
+        --timeout)
+            shift
+            TIMEOUT=$1
+            ;;
+        --email)
+            shift
+            EMAIL=$1
+            ;;
+        --efs-dns)
+            shift
+            EFS_DNS=$1
+            ;;
+        -*)
+            # do not exit out, just note failure
+            echo 1>&2 "unrecognized option: $1"
+            ;;
+        *)
+            break;
+            ;;
+    esac
+    shift
+done
+
+
+export ANACONDA_PATH=/mnt/anaconda2
+
+CURRENT_KEYS=`cat $HOME/.ssh/authorized_keys`
 TELEMETRY_CONF_BUCKET=s3://telemetry-spark-emr-2
 MEMORY_OVERHEAD=7000  # Tuned for c3.4xlarge
 EXECUTOR_MEMORY=15000M
 DRIVER_MIN_HEAP=1000M
 DRIVER_MEMORY=$EXECUTOR_MEMORY
+SETUP_HOME_DIR=false
+
+#EFS Mounting
+if [ -n "$EMAIL" ] && [ -n "$EFS_DNS" ] && [ "$IS_MASTER" = true ]; then
+    # Mount entire efs locally to check for user's dir
+    mkdir -p /mnt/efs
+    sudo mount -t nfs4 -o nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2 "$EFS_DNS:/" /mnt/efs
+
+    if [ ! -d "/mnt/efs/$EMAIL" ]; then
+        SETUP_HOME_DIR=true
+
+        # If no dir for this user, create it, create analyses dir, copy keys
+        sudo mkdir -p "/mnt/efs/$EMAIL/analyses"
+        sudo mkdir -p "/mnt/efs/$EMAIL/.ssh"
+        sudo cp "$HOME/.ssh/authorized_keys" "/mnt/efs/$EMAIL/.ssh"
+    fi
+
+    # mount user's EFS dir on $HOME
+    sudo mount -t nfs4 -o nfsvers=4.1,rsize=1048576,wsize=1048576,hard,timeo=600,retrans=2 "$EFS_DNS:/$EMAIL" "$HOME"
+
+    # Set ownership of user dir
+    sudo chown --recursive hadoop:hadoop "$HOME"
+
+    # allow Jupyter to access analyses dir
+    sudo chmod --recursive 777 "$HOME/analyses"
+
+    # fixes ssh access
+    sudo chmod go-w "$HOME"
+    sudo chmod 700 "$HOME/.ssh"
+    sudo chmod 600 "$HOME/.ssh/authorized_keys"
+elif [ "$IS_MASTER" = true ]; then
+    mkdir -p $HOME/analyses
+    SETUP_HOME_DIR=true
+fi
+
+if [ "$SETUP_HOME_DIR" = true ] ; then
+    # Only execute this when using EFS and this is a new user, or when not using EFS
+
+    # Download examples to analyses dir
```

Comment:
This is a discussion we had in the previous PR. I don't want to change *anything* in a user's dir, so they are just going to be there to begin with, and never updated or overwritten. They do not change often. We can kick the can down the road on whether we will want to add more examples later, since we can always just go through and stick them in people's dirs if we feel we really need to.